### PR TITLE
Java Relationships Support

### PR DESCRIPTION
**Pull request by TanayParikh**
- Added Java Relationship Generation
- Added Java RelationshipDAO
- Added Java ManyToMany Relationship DAO

**NOTE:** The general structure of the ManyToMany Relationship DAO has been completed but it requires the implementation of Java multi-part keys, so I have set it up to throw an exception for now.

**NOTE:** Need a way to thoroughly test this functionality. Would appreciate input on how I could go about doing so.

### CODE REVIEW

File: src/foam/dao/Relationship.js [hunk 1]
```diff
@@ -74,13 +78,15 @@ foam.CLASS({
     {
       class: 'FObjectArray',
       name: 'sourceProperties',
-      of: 'Property',
```
Comment (TanayParikh) [inline src/foam/dao/Relationship.js:77]: `Property` changed to `PropertyInfo` for Java generation, could this cause issues on the JS side?

`javaType` alone won't suffice as a `PropertyInfo` array is needed:

`Relationship.java`, L:194 (cast function generated in property info) if `of` is set to `Property`:

```java
foam.core.PropertyInfo[] ret = new foam.core.Property[value.length];
```

Is this the intended behaviour?

File: src/foam/dao/ManyToManyRelationshipDAO.js [hunk 2]
```diff
@@ -20,19 +20,42 @@ foam.CLASS({
...
+      name: 'sourceKey'
+    },
+    {
+      class: 'foam.core.Property',
+      javaType: 'foam.core.PropertyInfo',
+      name: 'targetProperty'
+    },
+    {
+      class: 'foam.core.Property',
+      javaType: 'foam.core.PropertyInfo',
```
Comment (kirkeaton) [inline src/foam/dao/ManyToManyRelationshipDAO.js]: Can you move all the javaType declarations to the java refinement class instead?

File: src/foam/dao/ManyToManyRelationshipDAOJava.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      // if (results == null) return results;
+
+      // return getDelegate().find_(x, innerId);
+      `
+    },
+    {
+      name: 'select_',
+      javaCode:
+      `
+      throw new UnsupportedOperationException("Many To Many Relationships await the implementation of multi-part keys in Java");
```
Comment (kirkeaton) [inline src/foam/dao/ManyToManyRelationshipDAOJava.js:36]: Is multi-part keys something that's being worked on at the moment?

File: src/foam/dao/ManyToManyRelationshipDAO.js (hunk 2 above)
Comment (TanayParikh) [reply src/foam/dao/ManyToManyRelationshipDAO.js]: Thanks, that's a great point. Updated.

File: src/foam/dao/ManyToManyRelationshipDAOJava.js (hunk 3 above)
Comment (TanayParikh) [reply src/foam/dao/ManyToManyRelationshipDAOJava.js:36]: I think it is in the pipeline.

File: src/foam/dao/Relationship.js [hunk 4]
```diff
@@ -46,6 +46,7 @@ foam.CLASS({
     },
     {
       name: 'name',
+      class: 'foam.core.String',
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Just use 'String', defaults to 'foam.core.' package.

File: src/foam/dao/Relationship.js [hunk 5]
```diff
@@ -54,7 +55,10 @@ foam.CLASS({
       }
     },
     'forwardName',
-    'inverseName',
+    {
+      name: 'inverseName',
+      class: 'foam.core.String'
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: Here also.

File: src/foam/dao/Relationship.js [hunk 6]
```diff
@@ -74,13 +78,15 @@ foam.CLASS({
     {
       class: 'FObjectArray',
       name: 'sourceProperties',
-      of: 'Property',
+      of: 'foam.core.PropertyInfo', // TODO: Property changed to Property info for Javaa generation, could this cause issues on the JS side?
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: s/Javaa/Java/

File: src/foam/dao/Relationship.js [hunk 7]
```diff
@@ -74,13 +78,15 @@ foam.CLASS({
...
+      javaType: 'foam.core.PropertyInfo[]',
       adaptArrayElement: foam.core.Model.PROPERTIES.adaptArrayElement
     },
     {
       class: 'FObjectArray',
       name: 'targetProperties',
-      of: 'Property',
+      of: 'foam.core.PropertyInfo', // TODO: Property changed to Property info for Javaa generation, could this cause issues on the JS side?
```
Comment (kgrgreer) [inline src/foam/dao/Relationship.js]: s/Javaa/Java/

File: src/foam/dao/RelationshipDAO.js [hunk 8]
```diff
@@ -24,10 +24,13 @@ foam.CLASS({
 
   properties: [
     {
-      name: 'obj'
+      name: 'obj',
+      class: 'foam.core.FObjectProperty'
```
Comment (kgrgreer) [inline src/foam/dao/RelationshipDAO.js]: Remove foam.core.

File: src/foam/dao/RelationshipDAO.js [hunk 9]
```diff
@@ -24,10 +24,13 @@ foam.CLASS({
...
     },
     {
       name: 'relationship',
+      class: 'foam.core.FObjectProperty',
```
Comment (kgrgreer) [inline src/foam/dao/RelationshipDAO.js]: Remove foam.core.

File: src/foam/dao/RelationshipDAOJava.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        }
+      ],
+      javaCode: 
+      `
+      String inverseName = getRelationship().getInverseName();
+
+      ((RelationshipDAO) target).getRelationship().setInverseName(source.getClassInfo().getId());
+
+      return (foam.core.FObject) target;
+      `
```
Comment (kgrgreer) [inline src/foam/dao/RelationshipDAOJava.js]: Remove blank line.

File: src/foam/java/refinements.js [hunk 11]
```diff
@@ -692,6 +692,12 @@ foam.CLASS({
 
 
 foam.CLASS({
+  refines: 'foam.core.Property',
+  javaType: 'foam.core.PropertyInfo'
```
Comment (kgrgreer) [inline src/foam/java/refinements.js]: Document why/where this is needed.

**Comment by kgrgreer** [comment]
Can you write a demo app to show how this is used?

File: src/foam/java/refinements.js (hunk 11 above)
Comment (TanayParikh) [reply src/foam/java/refinements.js]: Done.

File: src/foam/dao/RelationshipDAOJava.js (hunk 10 above)
Comment (TanayParikh) [reply src/foam/dao/RelationshipDAOJava.js]: Removed.

File: src/foam/dao/RelationshipDAO.js (hunk 9 above)
Comment (TanayParikh) [reply src/foam/dao/RelationshipDAO.js]: Removed.

File: src/foam/dao/RelationshipDAO.js (hunk 8 above)
Comment (TanayParikh) [reply src/foam/dao/RelationshipDAO.js]: Removed

File: src/foam/dao/Relationship.js (hunk 7 above)
Comment (TanayParikh) [reply src/foam/dao/Relationship.js]: Done.

File: src/foam/dao/Relationship.js (hunk 6 above)
Comment (TanayParikh) [reply src/foam/dao/Relationship.js]: Done.

File: src/foam/dao/Relationship.js (hunk 5 above)
Comment (TanayParikh) [reply src/foam/dao/Relationship.js]: Done.

File: src/foam/dao/Relationship.js (hunk 4 above)
Comment (TanayParikh) [reply src/foam/dao/Relationship.js]: Done.

**Comment by TanayParikh** [comment]
@kgrgreer That's just it, I'm not too sure what an effective way to test this functionality would be.  Would appreciate input on how I could go about doing so. I think once we integrate it within the existing project, it will provide a good environment to test this out (via the project as a proxy).

**Pull request merged**